### PR TITLE
Allow to customize response body of 403 issued by HTTP policy

### LINKED ISSUE #5751: Allow quarkus-oidc to customize response body of 403
**Description**
When using quarkus-oidc, providing a wrong token (expired, wrong signature, realm, ...) returns a forbidden status, as expected. However, the response body is always empty.

In order to migrate an existing API (which already defines its own response bodies), we would require from quarkus-oidc the possibility to customize and give more information in the response body.

For example, an expired token could produce this response body:
```
{
   "reason": "token_expired",
   "message": "Your token is expired."
}
```

### PR DESCRIPTION
**Pull request by michalvavrik**
closes: #5751

Allow to customize response body of 403 that comes from HTTP policy by exception mappers. IMO the rest of #5751 has already been fixed as explain in https://github.com/quarkusio/quarkus/issues/5751#issuecomment-1329673303. This PR throws `ForbiddenException` rather simply failing event with 403, that allows builtin exception mappers that are present in both RESTEasy Reactive and Classic to handle the exception.

### CODE REVIEW

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java [hunk 1]
```diff
@@ -218,16 +219,10 @@ public Handler<RoutingContext> failureHandler(RuntimeValue<RestInitialHandler> r
             @Override
             public void handle(RoutingContext event) {
 
-                // this condition prevent exception mappers from handling auth failure exceptions when proactive
```
Comment (sberyozkin) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java:221]: @michalvavrik Why have you removed this check ? Is it because you'd like to handle `403` ?

Comment (michalvavrik) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java:221]: @sberyozkin I mentioned I forgot to add same check in the RESTEasy Classic, therefore it already can handle authentication failures with exception mappers, while RESTEasy Reactive must use failure handler when `quarkus.http.auth.proactive=true`. It is important that user experience is same for both classic and reactive. I'm sorry I didn't mention this before.

I offer following options:

1. keep it as is. IMO it's good for users as typical RESTEasy Reactive user is used to exception mappers rather than failure handlers
2. I'll adjust RESTEasy Classic so that users can't use exception mappers with `quarkus.http.auth.proactive=true`

Comment (michalvavrik) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java:221]: > @michalvavrik Why have you removed this check ? Is it because you'd like to handle `403` ?

This PR only adds same treatment for ForbiddenException as we have for AuthenticationFailedE... etc. And also fails with exception so that it can be caught. The reason for removing this check is explained above.

**Comment by michalvavrik** [comment]
`io.quarkus.smallrye.reactivemessaging.amqp.AnonymousAmqpTest` seems like flaky test to me. I don't think it has relation to this PR.

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java (hunk 1 above)
Comment (sberyozkin) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java:221]: @michalvavrik Ok, so the reason I was asking for the filter only solution for the proactive auth case was 1) to avoid the concerns about the failed authentication be handled at the JAX-RS chain level, past the security layer - yes, we discussed it with Stuart, I suggested it was not a problem, but then we agreed to do it in your earlier PR anyway. 2) To give users an an easy way to opt out of having to map the error if for ex, Basic and OIDC are used and only a basic mapping is supported in the custom mapping code as mapping correctly for ex OIDC redirect exceptions may not be possible.

I agree we have to keep it consistent, and the fact it was not done for ResteasyClassic makes it difficult to maneuver. 
What happens now in 2.15 if ResteasyReactive users will try to handle the exception with the fault route, as per the docs removed in this PR ? And how ResteasyReactive users will be able to handle a case where both Basic and OIDC exceptions can be thrown but they only want to map a Basic AuthenticationFailedException ?

I think this PR has to be split into 2 PRs, the first one to deal with 403 only, the next one aligning Resteasy Classic and Reactive approaches - or choosing not to. Perhaps we need a 2nd PR prepared with a discussion in the issue first

thanks

Comment (michalvavrik) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java:221]: > I agree we have to keep it consistent, and the fact it was not done for ResteasyClassic makes it difficult to maneuver. What happens now in 2.15 if ResteasyReactive users will try to handle the exception with the fault route, as per the docs removed in this PR ?

It will work anyway (with and without this PR) and if users follow documentation, RESTEasy Classic users can and will do it to, so no problem (though I'm sorry for troubles!).

> And how ResteasyReactive users will be able to handle a case where both Basic and OIDC exceptions can be thrown but they only want to map a Basic AuthenticationFailedException ?

I actually couldn't find scenario where OIDC exception is not _finally_ transformed authenticationfailed/redirect/completion exception. Sometimes it's hard to follow flow, so I can be wrong. Anyway, only AuthenticationFailedException/AuthenticationCompletionException/AuthenticationRedirectException get to JAX-RS, rest is handled exclusively by failure route. I believe previous or current changes don't have impact on other exceptions.

> I think this PR has to be split into 2 PRs, the first one to deal with 403 only, the next one aligning Resteasy Classic and Reactive approaches - or choosing not to. Perhaps we need a 2nd PR prepared with a discussion in the issue first

Alright, let me to handle 403 exclusively in this PR and once it is done, I can propose changes that will level behavior in whichever way (failure handlers /ex. mappers).

Comment (michalvavrik) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java:221]: just fyi - you say correctly failed auth is handled at the JAX-RS chain level for RESTEasy Classic, but it is also important to know goes directly to the exception mappers, personally it doesn't worry me.

Comment (sberyozkin) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java:221]: @michalvavrik 

>It will work anyway (with and without this PR) and if users follow documentation, RESTEasy Classic users can and will do it too, so no problem (though I'm sorry for troubles!).

Np, you were doing a lot of fixes and unfortunately sometimes we can miss something.
If you say it will work anyway then we'll need to keep the example showing how to use the failure handler

>I actually couldn't find scenario where OIDC exception is not finally transformed authenticationfailed/redirect/completion exception. Sometimes it's hard to follow flow, so I can be wrong. Anyway, only AuthenticationFailedException/AuthenticationCompletionException/AuthenticationRedirectException get to JAX-RS, rest is handled exclusively by failure route.

Right, the users will want to handle `AuthenticationFailedException`, ex, for Basic - add some extra test in the response. Now the same mapper gets OIDC `AuthenticationFailedException` - this is supposed to be handled by OIDC to create a correct challenge, prepare a state cookie, etc, the custom mapper will most likely not want to deal with it. So with the failure handler I can imagine that for ex RoutingContext can be changed which mechanism was involved and opt out easily if it is OIDC. The exception mapper might be able to check the context - but how to opt out is not clear. 

>Alright, let me to handle 403 exclusively in this PR and once it is done, I can propose changes that will level behavior in whichever way (failure handlers /ex. mappers).

Yeah, IMHO it will make it simpler to discuss

>just fyi - you say correctly failed auth is handled at the JAX-RS chain level for RESTEasy Classic, but it is also important to know it goes directly to the exception mappers, personally it doesn't worry me.

Do you mean it is of no real concern if the mapper is called in case of the proactive auth ? I think I'm not too worried either, but lets talk about it in the follow up issue/PR

Thanks

Comment (michalvavrik) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java]: Done, the PR now handlers just 403.

> Perhaps we need a 2nd PR prepared with a discussion in the issue first

This is bit problem as second PR must be build on this one - f.e. adding ForbiddenException as I did in ResteasyReactiveRecord even with the condition would make it mappers work as RR failure handler is added to the event between authentication and permission handler. Best I can do is to open an issue.

Comment (michalvavrik) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java]: I'd suggest I simply add a PR a day after this one gets merged as I've research and tests already, but feel free to correct me.

Comment (michalvavrik) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java]: > Right, the users will want to handle AuthenticationFailedException, ex, for Basic - add some extra test in the response. Now the same mapper gets OIDC AuthenticationFailedException - this is supposed to be handled by OIDC to create a correct challenge, prepare a state cookie, etc, the custom mapper will most likely not want to deal with it. So with the failure handler I can imagine that for ex RoutingContext can be checked which mechanism was involved and opt out easily if it is OIDC. The exception mapper might be able to check the context - but how to opt out is not clear.

Builtin `AuthenticationFailedException` provided by RESTEasy classic and reactive get challenge, so the issue is as you say if user defines custom exception mapper and don't know what he do. We can discuss it in dedicated PR. I'll think how to do it.

Comment (michalvavrik) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java]: > Right, the users will want to handle AuthenticationFailedException, ex, for Basic - add some extra test in the response. Now the same mapper gets OIDC AuthenticationFailedException - this is supposed to be handled by OIDC to create a correct challenge, prepare a state cookie, etc, the custom mapper will most likely not want to deal with it. So with the failure handler I can imagine that for ex RoutingContext can be checked which mechanism was involved and opt out easily if it is OIDC. The exception mapper might be able to check the context - but how to opt out is not clear.

Perhaps SPI can contain constant SKIP_EXCEPTION_MAPPER that OIDC will add and RESTEasy Reactive/Classic won't even start processing in such cases.

I think users can already define these mappers that interfere with OIDC challenge when proactive=false.

Comment (sberyozkin) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java]: Right, just not sure adding such constants is the right way to handle it though...

**Review by sberyozkin — APPROVED** [review]
